src/OSPSuite.SimModelNative/src/Simulation.cpp

```diff
@@ -93,7 +93,19 @@ void Simulation::Finalize ()
 	for (int i = 0; i < _parameters.size(); i++)
 	{
 		if (_parameters[i]->CalculateSensitivity())
+		{
 			_sensitivityParameters.Add(_parameters[i]);
+			//cache sensitivity derivatives
+			for (int speciesIndex = 0; speciesIndex < m_ODE_NumUnknowns; speciesIndex++)
+			{
+				Species* species = GetDEVariableFromIndex(speciesIndex);
+				species->CalculateJacobianParameterFor(i);
+				for (int otherSpeciesIndex = 0; otherSpeciesIndex < m_ODE_NumUnknowns; otherSpeciesIndex++)
+				{
+					species->CalculateJacobianStateVariableFor(otherSpeciesIndex);
+				}
+			}
+		}
```

Comment:
This is done to early here. In the next steps ODE system is modified and ODE variables indices might change.
The earliest stage where this can be done is after 
```
// Second pass: Determine equation numbers
DE_SetSpeciesIndex();	
```

And I would do it even later, after the step 
`SimulationTask::CacheRHSUsedVariables(this);`

The step above caches for each ODE variable the indices of all ODE variables used in the RHS equation. And only for those variables jacobian would be not null. 
So I would add a new function to the `Species `class, something like **CacheJacobianFunctions** (feel free to find a better name :)) which:
* Runs through the indices of used variables and adds their jacobian function into the map
* For all other (unused) ODE Variables: add a constant zero formula (this should be the case anyway if adding a jacobian for such a variable: so you could use it as a test at the beginning :))

Reply:
Actually, it will calculate the derivative on its first use if not already calculated, so I will just remove the call to the cache since it will be created on the first iteration when running the simulation anyways.